File: packages/clients/android/src/main/kotlin/com/doubloon/checker/RpcClient.kt
package com.doubloon.checker

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicInteger

/**
 * Minimal JSON-RPC client using OkHttp.
 */
class RpcClient(
    private val client: OkHttpClient = OkHttpClient(),
) {
    private val idCounter = AtomicInteger(0)

    class RpcException(val code: Int, message: String) : Exception("RPC error $code: $message")

    suspend fun call(url: String, method: String, params: JSONArray): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("jsonrpc", "2.0")
                put("id", idCounter.incrementAndGet())
                put("method", method)
                put("params", params)
            }

            val request = Request.Builder()
                .url(url)
                .post(body.toString().toRequestBody("application/json".toMediaType()))
                .build()

            val response = client.newCall(request).execute()
            val responseBody = response.body?.string()
                ?: throw RpcException(-1, "Empty response body")

            if (!response.isSuccessful) {
                throw RpcException(response.code, "HTTP ${response.code}")
            }

            val json = JSONObject(responseBody)

            if (json.has("error")) {
                val error = json.getJSONObject("error")
                throw RpcException(
                    error.optInt("code", -1),
                    error.optString("message", "Unknown error")
                )
            }

            json
        }

    suspend fun callForString(url: String, method: String, params: JSONArray): String {
        val json = call(url, method, params)
        return json.getString("result")
    }
}

File: packages/clients/ios/Sources/DoubloonChecker/RpcClient.swift
import Foundation

/// Minimal JSON-RPC client using URLSession.
actor RpcClient {
    private let session: URLSession
    private var idCounter: Int = 0

    init(session: URLSession = .shared) {
        self.session = session
    }

    struct RpcError: Error, CustomStringConvertible {
        let code: Int
        let message: String
        var description: String { "RPC error \(code): \(message)" }
    }

    func call<T: Decodable>(url: URL, method: String, params: Any) async throws -> T {
        idCounter += 1
        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "id": idCounter,
            "method": method,
            "params": params,
        ]

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)

        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode >= 400 {
            throw RpcError(code: httpResponse.statusCode, message: "HTTP \(httpResponse.statusCode)")
        }

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]

        if let error = json["error"] as? [String: Any] {
            throw RpcError(
                code: error["code"] as? Int ?? -1,
                message: error["message"] as? String ?? "Unknown RPC error"
            )
        }

        // For raw string results (eth_call), return directly
        if T.self == String.self, let result = json["result"] as? String {
            return result as! T
        }

        // For complex results, re-serialize the "result" and decode
        let resultData = try JSONSerialization.data(withJSONObject: json["result"] ?? NSNull())
        return try JSONDecoder().decode(T.self, from: resultData)
    }
}
